File: scripts/codereview/py/call_graph.py
#!/usr/bin/env python3
"""
Python call graph analyzer using AST.

Analyzes Python files and builds function-level call relationships.
Outputs JSON for consumption by the Go wrapper.

Usage:
    python3 call_graph.py <files...> [--functions func1,func2,...]
"""

import ast
import json
import os
import sys
from dataclasses import dataclass, field
from collections.abc import Iterator
from typing import Optional

# Maximum file size to analyze (10MB) - prevents memory exhaustion
MAX_FILE_SIZE = 10 * 1024 * 1024


@dataclass
class CallSite:
    """Represents a call site within a function."""

    target: str
    line: int
    column: int
    is_method: bool


@dataclass
class CallerInfo:
    """Represents information about a function caller."""

    function: str
    file: str
    line: int


@dataclass
class FunctionInfo:
    """Comprehensive function information including call relationships."""

    name: str
    file: str
    line: int
    end_line: int
    call_sites: list[CallSite] = field(default_factory=list)
    called_by: list[CallerInfo] = field(default_factory=list)


@dataclass
class CallGraphOutput:
    """Output format for the call graph analysis."""

    functions: list[FunctionInfo] = field(default_factory=list)
    error: Optional[str] = None


class CallGraphVisitor(ast.NodeVisitor):
    """AST visitor that extracts function definitions and their call sites."""

    def __init__(self, file_path: str):
        self.file_path = file_path
        self.functions: list[FunctionInfo] = []
        self._class_stack: list[str] = []
        self._current_function: Optional[FunctionInfo] = None

    def visit_ClassDef(self, node: ast.ClassDef) -> None:
        """Visit a class definition and track class context for methods."""
        self._class_stack.append(node.name)
        self.generic_visit(node)
        self._class_stack.pop()

    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
        """Visit a function definition."""
        self._process_function(node)

    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef) -> None:
        """Visit an async function definition."""
        self._process_function(node)

    def _process_function(
        self, node: ast.FunctionDef | ast.AsyncFunctionDef
    ) -> None:
        """Process a function or async function definition."""
        # Build qualified name with class prefix if within a class
        if self._class_stack:
            name = f"{self._class_stack[-1]}.{node.name}"
        else:
            name = node.name

        func_info = FunctionInfo(
            name=name,
            file=self.file_path,
            line=node.lineno,
            end_line=node.end_lineno or node.lineno,
            call_sites=[],
            called_by=[],
        )

        # Store previous function context and set current
        prev_function = self._current_function
        self._current_function = func_info

        # Extract call sites from the function body
        # Use walk_without_nested_functions to prevent double-counting
        # calls that appear in nested functions
        call_extractor = CallSiteExtractor()
        for child in walk_without_nested_functions(node):
            if isinstance(child, ast.Call):
                call_site = call_extractor.extract_call(child)
                if call_site:
                    func_info.call_sites.append(call_site)

        self.functions.append(func_info)

        # Visit nested functions
        self.generic_visit(node)

        # Restore previous function context
        self._current_function = prev_function


def walk_without_nested_functions(node: ast.AST) -> Iterator[ast.AST]:
    """
    Walk AST nodes without descending into nested function definitions.

    This prevents double-counting calls that appear in nested functions.
    """
    from collections import deque
    todo = deque([node])
    while todo:
        current = todo.popleft()
        # Skip the root node itself if it's a function (we want its children)
        if current is not node and isinstance(
            current, (ast.FunctionDef, ast.AsyncFunctionDef)
        ):
            # Don't descend into nested functions
            continue
        yield current
        todo.extend(ast.iter_child_nodes(current))


class CallSiteExtractor:
    """Extracts call information from AST Call nodes."""

    def extract_call(self, node: ast.Call) -> Optional[CallSite]:
        """Extract call information from a Call node."""
        target = self._get_call_target(node.func)
        if not target:
            return None

        # Skip common built-ins that aren't useful for call graph
        skip_targets = {
            "print",
            "len",
            "str",
            "int",
            "float",
            "list",
            "dict",
            "set",
            "tuple",
            "range",
            "enumerate",
            "zip",
            "map",
            "filter",
            "sorted",
            "reversed",
            "isinstance",
            "issubclass",
            "hasattr",
            "getattr",
            "setattr",
            "delattr",
            "type",
            "super",
            "open",
            "input",
        }

        base_target = target.split(".")[-1]
        if base_target in skip_targets:
            return None

        is_method = isinstance(node.func, ast.Attribute)

        return CallSite(
            target=target,
            line=node.lineno,
            column=node.col_offset + 1,
            is_method=is_method,
        )

    def _get_call_target(self, func: ast.expr) -> Optional[str]:
        """Get the target name from a function call expression."""
        if isinstance(func, ast.Name):
            # Direct function call: function_name()
            return func.id
        elif isinstance(func, ast.Attribute):
            # Method call: obj.method() or module.function()
            value = self._get_expr_name(func.value)
            if value:
                return f"{value}.{func.attr}"
            return func.attr
        elif isinstance(func, ast.Subscript):
            # Subscript call: obj[key]()
            return "<subscript>"
        elif isinstance(func, ast.Call):
            # Chained call: func()()
            return "<chained>"
        return None

    def _get_expr_name(self, expr: ast.expr) -> Optional[str]:
        """Get a name representation for an expression."""
        if isinstance(expr, ast.Name):
            return expr.id
        elif isinstance(expr, ast.Attribute):
            value = self._get_expr_name(expr.value)
            if value:
                return f"{value}.{expr.attr}"
            return expr.attr
        elif expr.__class__.__name__ == "Constant":
            # ast.Constant for literals
            return None
        return None


def is_test_function(name: str) -> bool:
    """Check if a function name indicates a test function."""
    # Remove class prefix if present
    base_name = name.split(".")[-1]

    # Python unittest/pytest patterns
    if base_name.startswith("test_"):
        return True
    if base_name.startswith("Test"):
        return True
    if base_name.endswith("_test"):
        return True

    return False


def is_test_file(file_path: str) -> bool:
    """Check if a file path indicates a test file."""
    base = os.path.basename(file_path)

    # Common test file patterns
    if base.startswith("test_"):
        return True
    if base.endswith("_test.py"):
        return True
    if "tests/" in file_path or "/tests/" in file_path:
        return True
    if "test/" in file_path or "/test/" in file_path:
        return True

    return False


def analyze_file(file_path: str) -> list[FunctionInfo]:
    """Analyze a single Python file and extract function information."""
    # Check file size before reading to prevent memory exhaustion
    try:
        file_size = os.path.getsize(file_path)
        if file_size > MAX_FILE_SIZE:
            return []  # Skip files that are too large
    except OSError:
        return []

    try:
        with open(file_path, encoding="utf-8") as f:
            source = f.read()
    except OSError:
        return []

    try:
        tree = ast.parse(source, filename=file_path)
    except SyntaxError:
        return []

    visitor = CallGraphVisitor(file_path)
    visitor.visit(tree)

    return visitor.functions


def build_caller_map(
    all_functions: list[FunctionInfo],
) -> dict[str, list[CallerInfo]]:
    """Build a map of function names to their callers."""
    caller_map: dict[str, list[CallerInfo]] = {}

    # Initialize map for all functions
    for fn in all_functions:
        caller_map[fn.name] = []

    # Build caller relationships
    for caller_fn in all_functions:
        for call_site in caller_fn.call_sites:
            target_name = call_site.target
            base_name = target_name.split(".")[-1]

            # Try to match with known functions
            for target_fn in all_functions:
                target_base = target_fn.name.split(".")[-1]
                if (
                    target_fn.name == target_name
                    or target_fn.name == base_name
                    or target_base == base_name
                ):
                    if target_fn.name in caller_map:
                        caller_map[target_fn.name].append(
                            CallerInfo(
                                function=caller_fn.name,
                                file=caller_fn.file,
                                line=call_site.line,
                            )
                        )
                    break

    return caller_map


def analyze_call_graph(
    file_paths: list[str], target_functions: Optional[list[str]] = None
) -> CallGraphOutput:
    """
    Analyze Python files and build function-level call relationships.

    Args:
        file_paths: List of file paths to analyze
        target_functions: Optional list of function names to focus on

    Returns:
        CallGraphOutput with analysis results
    """
    result = CallGraphOutput()

    if not file_paths:
        return result

    # Filter out non-existent files
    existing_files = [fp for fp in file_paths if os.path.isfile(fp)]

    if not existing_files:
        result.error = "No valid files to analyze"
        return result

    # Collect all functions from all files
    all_functions: list[FunctionInfo] = []
    for file_path in existing_files:
        functions = analyze_file(file_path)
        all_functions.extend(functions)

    # Build caller map
    caller_map = build_caller_map(all_functions)

    # Filter functions if target list provided
    functions_to_report = all_functions
    if target_functions:
        target_set = set(target_functions)
        functions_to_report = [
            fn
            for fn in all_functions
            if fn.name in target_set or fn.name.split(".")[-1] in target_set
        ]

    # Build output with caller information
    for fn in functions_to_report:
        callers = caller_map.get(fn.name, [])
        fn.called_by = callers
        result.functions.append(fn)

    return result


def to_dict(obj) -> dict:
    """Convert a dataclass object to a dictionary for JSON serialization."""
    if hasattr(obj, "__dataclass_fields__"):
        result = {}
        for field_name in obj.__dataclass_fields__:
            value = getattr(obj, field_name)
            if isinstance(value, list):
                result[field_name] = [to_dict(item) for item in value]
            elif value is None:
                continue  # Skip None values
            else:
                result[field_name] = value
        return result
    return obj


def output_to_json(output: CallGraphOutput) -> str:
    """Convert CallGraphOutput to JSON string."""
    data = {
        "functions": [to_dict(fn) for fn in output.functions],
    }
    if output.error:
        data["error"] = output.error
    return json.dumps(data, indent=2)


def parse_args(args: list[str]) -> tuple[list[str], list[str]]:
    """Parse CLI arguments."""
    files: list[str] = []
    functions: list[str] = []

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == "--functions" and i + 1 < len(args):
            # Parse comma-separated function names
            func_list = args[i + 1]
            functions.extend(
                f.strip() for f in func_list.split(",") if f.strip()
            )
            i += 2
        elif not arg.startswith("--"):
            files.append(arg)
            i += 1
        else:
            i += 1

    return files, functions


def main() -> None:
    """Main CLI entry point."""
    args = sys.argv[1:]

    if not args:
        print(
            "Usage: call_graph.py <files...> [--functions func1,func2,...]",
            file=sys.stderr,
        )
        print("", file=sys.stderr)
        print(
            "Analyzes Python files and outputs function call relationships.",
            file=sys.stderr,
        )
        print("", file=sys.stderr)
        print("Options:", file=sys.stderr)
        print(
            "  --functions    Comma-separated list of function names to focus on",
            file=sys.stderr,
        )
        sys.exit(1)

    files, functions = parse_args(args)

    if not files:
        print("Error: No files specified", file=sys.stderr)
        sys.exit(1)

    # Resolve file paths
    resolved_files = [os.path.abspath(f) for f in files]

    try:
        target_functions = functions if functions else None
        result = analyze_call_graph(resolved_files, target_functions)
        print(output_to_json(result))
    except Exception as e:
        output = CallGraphOutput(error=f"{type(e).__name__}: {e}")
        print(output_to_json(output))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/codereview/ts/call-graph.ts
import * as fs from "fs"
import * as path from "path"
import * as ts from "typescript"

const IGNORED_CALLEE_TARGETS = new Set<string>([
  // Common Node/TS/JS built-ins / globals that usually add noise to call graphs.
  "require",
  "console.log",
  "console.error",
  "console.warn",
  "console.info",
  "console.debug",
  "JSON.stringify",
  "JSON.parse",
  "Object.keys",
  "Object.values",
  "Object.entries",
])

/**
 * Represents a call site within a function.
 */
interface CallSite {
  target: string
  line: number
  column: number
  is_method: boolean
}

/**
 * Represents information about a function caller.
 */
interface CallerInfo {
  function: string
  file: string
  line: number
}

/**
 * Represents comprehensive function information including call relationships.
 */
interface FunctionInfo {
  name: string
  file: string
  line: number
  end_line: number
  call_sites: CallSite[]
  called_by: CallerInfo[]
}

/**
 * Output format for the call graph analysis.
 */
interface CallGraphOutput {
  functions: FunctionInfo[]
  error?: string
}

/**
 * Internal representation of a parsed function during analysis.
 */
interface ParsedFunction {
  name: string
  file: string
  startLine: number
  endLine: number
  callSites: CallSite[]
  node: ts.Node
}

/**
 * Analyzes TypeScript files and builds function-level call relationships.
 *
 * @param filePaths - Array of file paths to analyze
 * @param targetFunctions - Optional array of function names to focus on
 * @returns Call graph analysis output
 */
function analyzeCallGraph(filePaths: string[], targetFunctions?: string[]): CallGraphOutput {
  const result: CallGraphOutput = {
    functions: [],
  }

  if (filePaths.length === 0) {
    return result
  }

  // Filter out non-existent files
  const existingFiles = filePaths.filter((fp) => {
    try {
      return fs.existsSync(fp)
    } catch {
      return false
    }
  })

  if (existingFiles.length === 0) {
    result.error = "No valid files to analyze"
    return result
  }

  const useTypeChecker = !(
    process.env.CALL_GRAPH_LIGHT_MODE === "1" || process.env.CALL_GRAPH_LIGHT_MODE === "true"
  )
  let typeChecker: ts.TypeChecker | null
  const fileSourceMap = new Map<string, ts.SourceFile>()
  const allFunctions: ParsedFunction[] = []

  if (useTypeChecker) {
    const program = ts.createProgram(existingFiles, {
      target: ts.ScriptTarget.Latest,
      module: ts.ModuleKind.CommonJS,
      allowJs: true,
      checkJs: true,
    })
    const programTypeChecker = program.getTypeChecker()
    typeChecker = programTypeChecker

    for (const filePath of existingFiles) {
      const sourceFile = program.getSourceFile(filePath)
      if (!sourceFile) {
        continue
      }
      fileSourceMap.set(filePath, sourceFile)
    }
  } else {
    typeChecker = null
    for (const filePath of existingFiles) {
      try {
        const content = fs.readFileSync(filePath, "utf-8")
        const sourceFile = ts.createSourceFile(filePath, content, ts.ScriptTarget.Latest, true)
        fileSourceMap.set(filePath, sourceFile)
      } catch {}
    }
  }

  if (fileSourceMap.size === 0) {
    result.error = "No readable files to analyze"
    return result
  }

  if (!useTypeChecker) {
    for (const [filePath, sourceFile] of fileSourceMap.entries()) {
      const functions = extractFunctionsLightMode(sourceFile, filePath)
      allFunctions.push(...functions)
    }
  } else {
    if (!typeChecker) {
      result.error = "Type checker unavailable"
      return result
    }

    const activeTypeChecker = typeChecker
    for (const [filePath, sourceFile] of fileSourceMap.entries()) {
      const functions = extractFunctions(sourceFile, filePath, activeTypeChecker)
      allFunctions.push(...functions)
    }
  }

  // Build caller map: function name -> callers
  const callerMap = buildCallerMap(allFunctions, existingFiles)

  // Filter functions if target list provided
  let functionsToReport = allFunctions
  if (targetFunctions && targetFunctions.length > 0) {
    const targetSet = new Set(targetFunctions)
    functionsToReport = allFunctions.filter(
      (fn) => targetSet.has(fn.name) || targetSet.has(getBaseName(fn.name)),
    )
  }

  // Build output
  for (const fn of functionsToReport) {
    const callers = callerMap.get(fn.name) || []

    result.functions.push({
      name: fn.name,
      file: fn.file,
      line: fn.startLine,
      end_line: fn.endLine,
      call_sites: fn.callSites,
      called_by: callers,
    })
  }

  return result
}

function extractFunctionsLightMode(sourceFile: ts.SourceFile, filePath: string): ParsedFunction[] {
  const functions: ParsedFunction[] = []

  function getLineNumber(pos: number): number {
    return sourceFile.getLineAndCharacterOfPosition(pos).line + 1
  }

  function extractCallSites(body: ts.Node): CallSite[] {
    const calls: CallSite[] = []

    function visitCalls(node: ts.Node) {
      if (ts.isCallExpression(node)) {
        calls.push({
          target: node.expression.getText(sourceFile),
          line: getLineNumber(node.getStart(sourceFile)),
          column: 1,
          is_method:
            ts.isPropertyAccessExpression(node.expression) ||
            ts.isElementAccessExpression(node.expression),
        })
      }

      if (ts.isNewExpression(node)) {
        calls.push({
          target: node.expression.getText(sourceFile),
          line: getLineNumber(node.getStart(sourceFile)),
          column: 1,
          is_method: false,
        })
      }

      ts.forEachChild(node, visitCalls)
    }

    visitCalls(body)
    return calls
  }

  function visit(node: ts.Node) {
    if (ts.isFunctionDeclaration(node) && node.name) {
      const callSites = node.body ? extractCallSites(node.body) : []
      functions.push({
        name: node.name.text,
        file: filePath,
        startLine: getLineNumber(node.getStart(sourceFile)),
        endLine: getLineNumber(node.getEnd()),
        callSites,
        node,
      })
    }

    if (ts.isVariableStatement(node)) {
      node.declarationList.declarations.forEach((decl) => {
        if (ts.isIdentifier(decl.name) && decl.initializer) {
          if (ts.isArrowFunction(decl.initializer) || ts.isFunctionExpression(decl.initializer)) {
            const funcExpr = decl.initializer
            const callSites = funcExpr.body ? extractCallSites(funcExpr.body) : []
            functions.push({
              name: decl.name.text,
              file: filePath,
              startLine: getLineNumber(node.getStart(sourceFile)),
              endLine: getLineNumber(node.getEnd()),
              callSites,
              node,
            })
          }
        }
      })
    }

    if (ts.isClassDeclaration(node) && node.name) {
      const className = node.name.text
      node.members.forEach((member) => {
        if (ts.isMethodDeclaration(member) && member.name) {
          const methodName = member.name.getText(sourceFile)
          const fullName = `${className}.${methodName}`
          const callSites = member.body ? extractCallSites(member.body) : []
          functions.push({
            name: fullName,
            file: filePath,
            startLine: getLineNumber(member.getStart(sourceFile)),
            endLine: getLineNumber(member.getEnd()),
            callSites,
            node: member,
          })
        }
      })
    }

    ts.forEachChild(node, visit)
  }

  visit(sourceFile)
  return functions
}

/**
 * Extracts all functions from a source file with their call sites.
 */
function extractFunctions(
  sourceFile: ts.SourceFile,
  filePath: string,
  typeChecker: ts.TypeChecker,
): ParsedFunction[] {
  const functions: ParsedFunction[] = []

  function extractCallSites(body: ts.Node): CallSite[] {
    const calls: CallSite[] = []

    function visitCalls(node: ts.Node) {
      if (ts.isCallExpression(node)) {
        const call = extractCallInfo(node, sourceFile, typeChecker)
        if (call) {
          calls.push(call)
        }
      }

      if (ts.isNewExpression(node)) {
        const call = extractNewExpressionInfo(node, sourceFile, typeChecker)
        if (call) {
          calls.push(call)
        }
      }

      ts.forEachChild(node, visitCalls)
    }

    visitCalls(body)
    return calls
  }

  function getLineNumber(pos: number): number {
    return sourceFile.getLineAndCharacterOfPosition(pos).line + 1
  }

  function visit(node: ts.Node) {
    // Function declarations
    if (ts.isFunctionDeclaration(node) && node.name) {
      const callSites = node.body ? extractCallSites(node.body) : []
      functions.push({
        name: node.name.text,
        file: filePath,
        startLine: getLineNumber(node.getStart(sourceFile)),
        endLine: getLineNumber(node.getEnd()),
        callSites,
        node,
      })
    }

    // Arrow functions assigned to const/let
    if (ts.isVariableStatement(node)) {
      node.declarationList.declarations.forEach((decl) => {
        if (ts.isIdentifier(decl.name) && decl.initializer) {
          if (ts.isArrowFunction(decl.initializer) || ts.isFunctionExpression(decl.initializer)) {
            const funcExpr = decl.initializer
            const callSites = funcExpr.body ? extractCallSites(funcExpr.body) : []
            functions.push({
              name: decl.name.text,
              file: filePath,
              startLine: getLineNumber(node.getStart(sourceFile)),
              endLine: getLineNumber(node.getEnd()),
              callSites,
              node,
            })
          }
        }
      })
    }

    // Class methods
    if (ts.isClassDeclaration(node) && node.name) {
      const className = node.name.text

      node.members.forEach((member) => {
        if (ts.isMethodDeclaration(member) && member.name) {
          const methodName = member.name.getText(sourceFile)
          const fullName = `${className}.${methodName}`
          const callSites = member.body ? extractCallSites(member.body) : []

          functions.push({
            name: fullName,
            file: filePath,
            startLine: getLineNumber(member.getStart(sourceFile)),
            endLine: getLineNumber(member.getEnd()),
            callSites,
            node: member,
          })
        }

        // Constructors
        if (ts.isConstructorDeclaration(member)) {
          const fullName = `${className}.constructor`
          const callSites = member.body ? extractCallSites(member.body) : []

          functions.push({
            name: fullName,
            file: filePath,
            startLine: getLineNumber(member.getStart(sourceFile)),
            endLine: getLineNumber(member.getEnd()),
            callSites,
            node: member,
          })
        }
      })
    }

    // Object literal methods (e.g., in module.exports)
    if (ts.isPropertyAssignment(node)) {
      if (
        node.initializer &&
        (ts.isFunctionExpression(node.initializer) || ts.isArrowFunction(node.initializer))
      ) {
        const funcExpr = node.initializer
        const name = node.name.getText(sourceFile)
        const callSites = funcExpr.body ? extractCallSites(funcExpr.body) : []

        functions.push({
          name,
          file: filePath,
          startLine: getLineNumber(node.getStart(sourceFile)),
          endLine: getLineNumber(node.getEnd()),
          callSites,
          node,
        })
      }
    }

    ts.forEachChild(node, visit)
  }

  visit(sourceFile)
  return functions
}

/**
 * Extracts call information from a CallExpression node.
 */
function extractCallInfo(
  node: ts.CallExpression,
  sourceFile: ts.SourceFile,
  typeChecker: ts.TypeChecker,
): CallSite | null {
  const { line, character } = sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile))

  function shouldIgnoreCall(): boolean {
    // `require(...)`
    if (ts.isIdentifier(node.expression) && node.expression.text === "require") {
      return true
    }

    // `console.*(...)`, `JSON.*(...)`, `Object.*(...)`
    const isIgnoredPropertyCall = (objName: string, memberName: string): boolean =>
      IGNORED_CALLEE_TARGETS.has(`${objName}.${memberName}`)

    if (ts.isPropertyAccessExpression(node.expression)) {
      const objExpr = node.expression.expression
      const memberName = node.expression.name.text
      if (ts.isIdentifier(objExpr) && isIgnoredPropertyCall(objExpr.text, memberName)) {
        return true
      }
    }

    // `console['log'](...)` and similar
    if (ts.isElementAccessExpression(node.expression)) {
      const objExpr = node.expression.expression
      const arg = node.expression.argumentExpression
      if (
        ts.isIdentifier(objExpr) &&
        arg &&
        ts.isStringLiteralLike(arg) &&
        isIgnoredPropertyCall(objExpr.text, arg.text)
      ) {
        return true
      }
    }

    return false
  }

  if (shouldIgnoreCall()) {
    return null
  }

  const isMethod =
    ts.isPropertyAccessExpression(node.expression) || ts.isElementAccessExpression(node.expression)

  function resolveCalledSymbol(): ts.Symbol | undefined {
    // Prefer resolving the symbol at the callee expression (usually yields the most human-meaningful symbol,
    // e.g. the variable/function name instead of an internal call-signature symbol like `__call`).
    let symbolFromLocation: ts.Symbol | undefined
    if (ts.isPropertyAccessExpression(node.expression)) {
      symbolFromLocation =
        typeChecker.getSymbolAtLocation(node.expression.name) ??
        typeChecker.getSymbolAtLocation(node.expression)
    } else if (ts.isElementAccessExpression(node.expression)) {
      const arg = node.expression.argumentExpression
      if (arg && ts.isStringLiteralLike(arg)) {
        symbolFromLocation =
          typeChecker.getSymbolAtLocation(arg) ?? typeChecker.getSymbolAtLocation(node.expression)
      } else {
        symbolFromLocation = typeChecker.getSymbolAtLocation(node.expression)
      }
    } else {
      symbolFromLocation = typeChecker.getSymbolAtLocation(node.expression)
    }

    // Also try resolved signature (useful for some call-signature cases where location resolution is missing).
    const signature = typeChecker.getResolvedSignature(node)
    if (signature) {
      const sigAny = signature as unknown as {
        getDeclaration?: () => ts.SignatureDeclaration | undefined
        declaration?: ts.SignatureDeclaration
      }
      const sigDecl = sigAny.getDeclaration ? sigAny.getDeclaration() : sigAny.declaration
      const sigSymbol = (sigDecl as unknown as { symbol?: ts.Symbol } | undefined)?.symbol
      if (!symbolFromLocation && sigSymbol) {
        return sigSymbol
      }
    }

    return symbolFromLocation
  }

  function getCanonicalTargetFromSymbol(sym: ts.Symbol): string | null {
    let symbol = sym
    if (symbol.flags & ts.SymbolFlags.Alias) {
      symbol = typeChecker.getAliasedSymbol(symbol)
    }

    const symbolName = symbol.getName()
    const declarations = symbol.getDeclarations() ?? []
    if (declarations.length === 0 && !symbolName) {
      return null
    }

    // Prefer a stable container-qualified name where possible (e.g., `ClassName.method`).
    for (const decl of declarations) {
      if (ts.isConstructorDeclaration(decl)) {
        const parent = decl.parent
        if ((ts.isClassDeclaration(parent) || ts.isClassExpression(parent)) && parent.name) {
          return `${parent.name.text}.constructor`
        }
      }

      if (
        ts.isMethodDeclaration(decl) ||
        ts.isMethodSignature(decl) ||
        ts.isPropertyDeclaration(decl) ||
        ts.isPropertySignature(decl) ||
        ts.isGetAccessorDeclaration(decl) ||
        ts.isSetAccessorDeclaration(decl)
      ) {
        const parent = decl.parent
        if (
          (ts.isClassDeclaration(parent) ||
            ts.isClassExpression(parent) ||
            ts.isInterfaceDeclaration(parent)) &&
          parent.name
        ) {
          return `${parent.name.text}.${symbolName}`
        }
      }
    }

    // Otherwise, include a file prefix as metadata while keeping the function name suffix stable:
    // `someFile.fnName` so `getBaseName(...)` still yields `fnName`.
    if (declarations.length > 0) {
      const declFile = declarations[0].getSourceFile().fileName
      const fileBase = path.basename(declFile, path.extname(declFile))
      if (fileBase && symbolName) {
        return `${fileBase}.${symbolName}`
      }
    }

    return symbolName || null
  }

  const resolvedSymbol = resolveCalledSymbol()
  let target = resolvedSymbol ? getCanonicalTargetFromSymbol(resolvedSymbol) : null

  // Fallback: only use text-based extraction if the type checker couldn't resolve a symbol.
  if (!target) {
    if (ts.isPropertyAccessExpression(node.expression)) {
      target = node.expression.getText(sourceFile)
    } else if (ts.isIdentifier(node.expression)) {
      target = node.expression.text
    } else if (ts.isElementAccessExpression(node.expression)) {
      target = "<dynamic>"
    } else {
      target = "<anonymous>"
    }

    if (IGNORED_CALLEE_TARGETS.has(target)) {
      return null
    }
  }

  return {
    target,
    line: line + 1,
    column: character + 1,
    is_method: isMethod,
  }
}

/**
 * Extracts call information from a NewExpression node.
 */
function extractNewExpressionInfo(
  node: ts.NewExpression,
  sourceFile: ts.SourceFile,
  typeChecker: ts.TypeChecker,
): CallSite | null {
  const { line, character } = sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile))

  let target: string

  if (ts.isIdentifier(node.expression)) {
    target = `new ${node.expression.text}`
  } else if (ts.isPropertyAccessExpression(node.expression)) {
    target = `new ${node.expression.getText(sourceFile)}`
  } else {
    target = "new <anonymous>"
  }

  return {
    target,
    line: line + 1,
    column: character + 1,
    is_method: false,
  }
}

/**
 * Builds a map of function names to their callers.
 */
function buildCallerMap(functions: ParsedFunction[], files: string[]): Map<string, CallerInfo[]> {
  const callerMap = new Map<string, CallerInfo[]>()

  // Initialize map for all functions
  for (const fn of functions) {
    callerMap.set(fn.name, [])
  }

  // Build caller relationships
  for (const callerFn of functions) {
    for (const callSite of callerFn.callSites) {
      // Find the target function
      const targetName = callSite.target
      const baseName = getBaseName(targetName)

      // Try to match with known functions
      for (const targetFn of functions) {
        if (
          targetFn.name === targetName ||
          targetFn.name === baseName ||
          getBaseName(targetFn.name) === baseName
        ) {
          const callers = callerMap.get(targetFn.name) || []
          callers.push({
            function: callerFn.name,
            file: callerFn.file,
            line: callSite.line,
          })
          callerMap.set(targetFn.name, callers)
          break
        }
      }
    }
  }

  return callerMap
}

/**
 * Gets the base name of a function (without class prefix).
 */
function getBaseName(name: string): string {
  const parts = name.split(".")
  return parts[parts.length - 1]
}

/**
 * Parses CLI arguments.
 */
function parseArgs(args: string[]): { files: string[]; functions: string[] } {
  const files: string[] = []
  const functions: string[] = []

  for (let i = 0; i < args.length; i++) {
    const arg = args[i]
    if (arg === "--functions" && i + 1 < args.length) {
      // Parse comma-separated function names
      const funcList = args[i + 1]
      functions.push(
        ...funcList
          .split(",")
          .map((f) => f.trim())
          .filter((f) => f),
      )
      i++
    } else if (!arg.startsWith("--")) {
      files.push(arg)
    }
  }

  return { files, functions }
}

/**
 * Main CLI entry point.
 */
function main() {
  const args = process.argv.slice(2)

  if (args.length === 0) {
    console.error("Usage: call-graph <files...> [--functions func1,func2,...]")
    console.error("")
    console.error("Analyzes TypeScript files and outputs function call relationships.")
    console.error("")
    console.error("Options:")
    console.error("  --functions    Comma-separated list of function names to focus on")
    process.exit(1)
  }

  const { files, functions } = parseArgs(args)

  if (files.length === 0) {
    console.error("Error: No files specified")
    process.exit(1)
  }

  // Resolve file paths
  const resolvedFiles = files.map((f) => path.resolve(f))

  try {
    const targetFunctions = functions.length > 0 ? functions : undefined
    const result = analyzeCallGraph(resolvedFiles, targetFunctions)
    console.log(JSON.stringify(result, null, 2))
  } catch (error) {
    const output: CallGraphOutput = {
      functions: [],
      error: error instanceof Error ? error.message : String(error),
    }
    console.log(JSON.stringify(output, null, 2))
    process.exit(1)
  }
}

main()
